basic guesser features (#3753)
MDAnalysis/mdanalysis: package/doc/sphinx/source/documentation_pages/guesser_modules.rst

.. Contains the formatted docstrings from the guesser modules located in 'mdanalysis/package/MDAnalysis/guesser'

**************************

> guesser modules

Reviewer: ```suggestion
Guesser modules
```
Just for consistency with the other documentation pages

**************************

> This module contains the context-aware guessers, which are used by the :ref:`guess_TopologyAttributes <guess_TopologyAttributes>` API. Context-aware guessers' main purpose

Reviewer 2: Here too, the `:ref:` should be switched to the fully specified Universe.guess_TopologyAttributes `:meth:` mentioned above

is to be tailored guesser classes that target specific file format or force field (eg. PDB file format, or Martini forcefield).

> Having such guessers makes attribute guessing more accurate and reliable than having generic guessing methods that doesn't  fit all topologies.

Reviewer: ```suggestion
Having such guessers makes attribute guessing more accurate and reliable than having generic guessing methods that doesn't fit all topologies.
```

Example uses of guessers
------------------------

> Guessing using :ref:`guess_TopologyAttributes <guess_TopologyAttributes>` API

Reviewer 2: Here as well -- although it might look nicer to just have plain `guess_TopologyAttributes` instead of the Sphinx markup!

> Guessing using :meth:`~MDAnalysis.core.Universe.guess_TopologyAttributes` API

Reviewer 2: ```suggestion
Guessing using :meth:`~MDAnalysis.core.universe.Universe.guess_TopologyAttributes` API
```

Just adding the full path here.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Guessing can be done through the universe's :func:`guess_TopologyAttributes` as following::

Reviewer 2: As mentioned [above](https://github.com/MDAnalysis/mdanalysis/pull/3753/files#r1439962814), this path probably has to be fully specified to link properly.

> Guessing can be done through the universe's :func:`~MDAnalysis.core.Universe.guess_TopologyAttributes` as following::

Reviewer 2: ```suggestion
Guessing can be done through the universe's :meth:`~MDAnalysis.core.universe.Universe.guess_TopologyAttributes` as following::
```

Same as above

> Guessing can be done through the universe's :meth:`~MDAnalysis.core.universe.Universe.guess_TopologyAttrs` as following::

Reviewer: ```suggestion
Guessing can be done through the Universe's :meth:`~MDAnalysis.core.universe.Universe.guess_TopologyAttrs` as following::
```

  import MDAnalysis as mda
  from MDAnalysisTests.datafiles import PDB

  u = mda.Universe(PDB)
  print(hasattr(u.atoms, 'elements'))  # returns False
  u.guess_TopologyAttributes(to_guess=['elements'])
  print(u.atoms.elements) # print ['N' 'H' 'H' ... 'NA' 'NA' 'NA']

In the above example, we passed ``elements`` as the attribute we want to guess, and

> :func:`~MDAnalysis.core.Universe.guess_TopologyAttributes` guess then add it as a topology

Reviewer 2: ```suggestion
:meth:`~MDAnalysis.core.universe.Universe.guess_TopologyAttributes` guess then add it as a topology
```

attribute to the ``AtomGroup`` of the universe.

If the attribute already exist in the universe, passing the attribute of interest to the ``to_guess`` parameter will only fill the empty values of the attribute if any exists.
To override all the attribute values, you can pass the attribute to the ``force_guess`` parameter instead of the to_guess one as following::

   import MDAnalysis as mda
   from MDAnalysisTests.datafiles import PRM12
 
   u = mda.Universe(PRM12, context='default', to_guess=()) # types ['HW', 'OW', ..]

   u.guess_TopologyAttributes(force_guess=['types']) # types ['H', 'O', ..]

> N.B.: If you didn't pass any ``context`` to the API, it will use the :ref:`DefaultGuesser <DefaultGuesser>`

Reviewer 2: ```suggestion
N.B.: If you didn't pass any ``context`` to the API, it will use the :class:`~MDAnalysis.guesser.default_guesser.DefaultGuesser`
```

Here too, if we use a fully specified path to a class, there's no need to add a `:ref:` :-)
